File: src/lexer/lexer.cpp
#include "lexer/lexer.h"
#include "lexer/source-file.h"
#include "lexer/source-line.h"
#include "util.h"

Lexer::Lexer(Fault& fault, SourceFile& file)
: fault(fault), tokens(), file(file) {}

str Lexer::toString() const { return vectorToString(tokens); }

Token Lexer::newToken(Token::Type type, SourceLine& line) {
    Token token(line, type);
    tokens.push_back(token);
    return token;
}

Token Lexer::newToken(Token::Type type) {
    return newToken(type, *file.line);
}

void Lexer::makeTokens() {
    do makeToken(); while(file.line->string != "");
}

void Lexer::makeToken() {
    ignoreSpaces();
    char next = file.next();

    if(next == '\0' or in(next, Token::PUNC_SYMS)) return makePunctuator();
    if(next == '\\') return makeLineContinue();
    if(next == '\'') return makeChar();
    if(next == '"') return makeString();
    if(in(next, Token::S_NUM_SYMS)) return makeNumber();
    if(in(next, Token::OP_SYMS)) return makeOperator();
    if(in(next, Token::S_IDENTIFIER_SYMS)) return makeIdentifier();

    file.take(1);
    Token token = newToken(Token::NONE);

    throw fault.fail(token, fmt::format(
        "Unrecognized symbol '{}'",
        token.string
    ));
}

void Lexer::ignoreSpaces() {
    while(isspace(file.next())) {
        if(file.next() == '\n') { 
            makeNewline();
            continue;
        }

        file.take(1);
        file.line->ignore();
    }
}    

void Lexer::ignoreComment() {
    if(file.take(1) == "/") {
        file.take(-1, "", "\n");
        file.line->ignore();
        return;
    } 

    SourceLine& line = *file.line;

    while(file.next()) {
        if(file.next() != '*') {
            file.take(-1, "", "*");
            continue;
        }

        file.take(1);

        if(file.next() == '/') {
            file.take(1);
            file.line->ignore();
            return;
        }
    }

    Token token = newToken(Token::NONE, line);
    throw fault.fail(token, "Unterminated multiline comment");
}

void Lexer::makePunctuator() {
    file.take(1);
    newToken(Token::PUNC);
}

void Lexer::makeLineContinue() {
    file.take(1);

    while(isspace(file.next())) {
        if(file.take(1) == "\n") return file.line->ignore();
    }

    file.line->ignore();
    file.take(-1, "", "\n");

    Token token = newToken(Token::NONE);
    throw fault.fail(token, "Unexpected symbols after line continuation");
}

void Lexer::makeNewline() {
    SourceLine& line = *file.line;

    while(file.next() == '\n') {
        file.take(1);
    }
    
    newToken(Token::PUNC, line);
}

void Lexer::makeChar() {
    file.take(1);
    
    if(file.next() == '\\') { 
        file.take(1);

        if(file.next() == '\'') file.take(1);
        file.take(-1, "", "'");
    } else file.take(1);

    if(file.take(1) == "'") {
        newToken(Token::CHAR);
        return;
    }
        
    Token token = newToken(Token::NONE);
    throw fault.fail(token, "Unterminated character");
}

void Lexer::makeString() {
    file.take(1);

    while(file.next() != '\n') {
        str taken = file.take(1);
        if(taken == "\\") file.take(1);

        if(taken == "\"") {
            newToken(Token::STR);
            return;
        }
    }

    Token token = newToken(Token::NONE);
    throw fault.fail(token, "Unterminated string");
}

void Lexer::makeNumber() {
    if(file.take(1) == "." and not in(file.next(), str("0123456789"))) {
        Token::Type type = Token::OP;
        
        if(file.next() == '.') {
            type = Token::PUNC;
            file.take(1);
            if(file.next() == '.') file.take(1); 
        }
        
        newToken(type);
        return;
    }

    file.take(-1, Token::NUM_SYMS);
    newToken(Token::NUM);
}

void Lexer::makeOperator() {
    str string;
    str next;

    checkNextSym:
    next = string + file.next();

    for(str op : Token::OPERATORS) {
        if(isSubstr(next, op)) {
            string += file.take(1);
            goto checkNextSym;
        }
    }

    if(string == "/" and in(file.next(), str("/*"))) return ignoreComment();
    Token token = newToken(in(string, Token::PUNC_OPS)? Token::PUNC : Token::OP);
    if(string == "###") return makeCStatement();

    if(not in(string, Token::OPERATORS)) {
        throw fault.fail(token, fmt::format(
            "Unrecognized operator '{}'", 
            token.string));
    }
}

void Lexer::makeIdentifier() {
    str string = file.take(-1, Token::IDENTIFIER_SYMS);
    newToken(in(string, Token::KEYWORDS)? Token::KEYWORD : Token::IDENTIFIER);
}

void Lexer::makeCStatement() {
    file.take(-1, "", "\n");
    newToken(Token::C_TOKENS);
}

File: app/src/main/kotlin/sea/lexer/Lexer.kt
package sea.lexer

import sea.Faults
import sea.lexer.SourceFile
import sea.lexer.Token
import sea.lexer.TokenType

data class Lexer(val faults: Faults, val file: SourceFile) {
    val tokens = ArrayList<Token>()

    override fun toString(): String = tokens.joinToString(", ", "[", "]")

    fun newToken(type: TokenType = TokenType.NONE, line: SourceLine = file.line): Token {
        val token = Token(line, type)
        tokens.add(token)
        return token
    }

    fun makeTokens() {
        do makeToken() while(!file.atEnd)
    }

    fun makeToken() {
        ignoreSpaces()

        when(file.next) {
            '\\' -> return makeLineContinuation()
            '\u0000', in Token.PUNC_SYMBOLS -> return makePunctuator()
            '\'' -> return makeChar()
            '\u0022' -> return makeString()
            in Token.NUMBER_START_SYMBOLS -> return makeNumber()
            in Token.OPERATOR_SYMBOLS -> return makeOperator()
            in Token.IDENTIFIER_START_SYMBOLS -> return makeIdentifier()
        }

        file.take(1)
        val token = newToken()

        throw faults.fail(token, "Unrecognized symbol '${token.string}'")
    }

    fun ignoreSpaces() {
        while(file.next.isWhitespace()) {
            if(file.next == '\n') {
                makeNewline()
                continue
            }

            file.take(1)
            file.line.ignore()
        }
    }

    fun ignoreComment() {
        if(file.take(1) == "/") {
            file.take(until = "\n")
            return file.line.ignore()
        }

        val line = file.line

        while(file.next != '\u0000') {
            if(file.next != '*') {
                file.take(until = "*")
                continue
            }
            
            file.take(1)

            if(file.next == '/') {
                file.take(1)
                return file.line.ignore()
            }
        }

        throw faults.fail(newToken(line = line), "Unterminated multiline comment")
    }

    fun makeNewline() {
        val line = file.line

        while(file.next == '\n') {
            file.take(1)
        }

        newToken(TokenType.PUNC, line)
    }

    fun makeLineContinuation() {
        file.take(1)

        while(file.next.isWhitespace()) {
            if(file.take(1) == "\n") return file.line.ignore()
        }

        file.line.ignore()
        file.take(until = "\n")

        throw faults.fail(newToken(), "Unexpected symbols after line continuation character")
    }

    fun makePunctuator() {
        file.take(1)
        newToken(TokenType.PUNC)
    }

    fun makeChar() {
        val pattern = """'(\\[btnr'"\\$]|\\u[a-fA-F0-9]{4}|[^\\'])'""".toRegex()
        file.take(1)

        if(file.next == '\\') {
            file.take(1)

            if(file.next == '\'') file.take(1)
            file.take(until = "'")
        } else file.take(1)

        if(file.take(1) == "'") {
            val token = newToken(TokenType.CHAR)
            if(token.string.matches(pattern)) return

            faults.error(token, "Invalid character")
            return
        }

        throw faults.fail(newToken(), "Unterminated character")
    }

    fun makeString() {
        file.take(1)

        while(file.next != '\n') {
            val taken = file.take(1)
            if(taken == "\\") file.take(1)

            if(taken == "\"") {
                newToken(TokenType.STR)
                return
            }
        }

        throw faults.fail(newToken(), "Unterminated string")
    }

    fun makeNumber() {
        var pattern = """(\d+)|(\d*\.\d+)|(\d+\.\d*)|"""
        pattern += """(\d+b(([0-9A-Z]+)|([0-9A-Z]*\.[0-9A-Z]+)|([0-9A-Z]+\.[0-9A-Z]*)))"""

        if(file.take(1) == "." && !(file.next.isDigit())) {
            newToken(TokenType.OP)
            return
        }

        file.take(these = Token.NUMBER_SYMBOLS)
        val token = newToken(TokenType.NUM)

        if(!token.string.matches(pattern.toRegex())) throw faults.fail(token, "Invalid number")
        token.convertNumber(faults)
    }

    fun makeOperator() {
        var string = ""
        var next: String

        checkNextSym@ while(true) {
            next = string + file.next

            for(op in Token.OPERATORS) {
                if(next in op) {
                    string += file.take(1)
                    continue@checkNextSym
                }
            }

            break
        }

        if(string == "/" && file.next in "/*") return ignoreComment()
        val token = newToken(if(string in Token.PUNC_OPS) TokenType.PUNC else TokenType.OP)

        if(string !in Token.OPERATORS) {
            throw faults.fail(token, "Unrecognized operator '${token.string}'")
        }
    }

    fun makeIdentifier() {
        val string = file.take(these = Token.IDENTIFIER_SYMBOLS)
        newToken(if(string in Token.KEYWORDS) TokenType.KEYWORD else TokenType.IDENTIFIER)
    }
}
